GDScript: Assume freed object to be of any Object type

vnen opened the pull request:
Since we can't check the actual type, we assume the type is correct if the expected type is Object-derived, the same as what happen with null values.

* Fix #86609

Edit: added an exception for `is` to also allow it on freed objects, but in this case it's always `false`. Also added a few test cases.

* _Bugsquad edit:_ Fixes #101423.

vnen committed:
GDScript: Assume freed object to be of any Object type

Since we can't check the actual type, we assume the type is correct if
the expected type is Object-derived, the same as what happen with null
values.

Also allow the `is` check to be performed on a freed Object, but in this
case it will always return `false`, since the actual type cannot be
checked.

dalexeev commented:
Makes a lot of sense to me, but we should check other places as well. For example, the following hack is probably no longer needed, at least replacing `Freed Object` with `null` (see #81605 for details):

https://github.com/godotengine/godot/blob/107f2961ccfac179af7682eb5f6e7ea91e80040c/modules/gdscript/gdscript_lambda_callable.cpp#L95-L101

The old behavior (error when passing `Freed Object` as a parameter) was convenient, perhaps people rely on it:

![](https://github.com/godotengine/godot/assets/47700418/a344283e-a89a-4132-bf42-3f9a85024684)

It seems that the useful error will not appear with this PR, since it does not appear now if you remove the type hint from the function parameter. Probably a similar check should be done separately, since we check this in many cases:

<details>
<summary>Incomplete list</summary>

https://github.com/godotengine/godot/blob/107f2961ccfac179af7682eb5f6e7ea91e80040c/modules/gdscript/gdscript_utility_functions.cpp#L550

https://github.com/godotengine/godot/blob/107f2961ccfac179af7682eb5f6e7ea91e80040c/modules/gdscript/gdscript_utility_functions.cpp#L567

https://github.com/godotengine/godot/blob/107f2961ccfac179af7682eb5f6e7ea91e80040c/modules/gdscript/gdscript_vm.cpp#L825

https://github.com/godotengine/godot/blob/107f2961ccfac179af7682eb5f6e7ea91e80040c/modules/gdscript/gdscript_vm.cpp#L847

https://github.com/godotengine/godot/blob/107f2961ccfac179af7682eb5f6e7ea91e80040c/modules/gdscript/gdscript_vm.cpp#L1362

https://github.com/godotengine/godot/blob/107f2961ccfac179af7682eb5f6e7ea91e80040c/modules/gdscript/gdscript_vm.cpp#L1399

https://github.com/godotengine/godot/blob/107f2961ccfac179af7682eb5f6e7ea91e80040c/modules/gdscript/gdscript_vm.cpp#L2249

https://github.com/godotengine/godot/blob/107f2961ccfac179af7682eb5f6e7ea91e80040c/modules/gdscript/gdscript_vm.cpp#L2560

https://github.com/godotengine/godot/blob/107f2961ccfac179af7682eb5f6e7ea91e80040c/modules/gdscript/gdscript_vm.cpp#L2602

https://github.com/godotengine/godot/blob/107f2961ccfac179af7682eb5f6e7ea91e80040c/modules/gdscript/gdscript_vm.cpp#L2970

https://github.com/godotengine/godot/blob/107f2961ccfac179af7682eb5f6e7ea91e80040c/modules/gdscript/gdscript_vm.cpp#L3301

</details>

By the way, I noticed an error, `was_freed` is not used:

https://github.com/godotengine/godot/blob/107f2961ccfac179af7682eb5f6e7ea91e80040c/modules/gdscript/gdscript_vm.cpp#L1721-L1730

vnen commented:
What I think is that is if a value is immediately used it should give an error. For instance, passing it to a utility function should raise the error because the user does not have control of its contents, but passing to a user-made function (like in this PR) should not trigger it automatically. Because sometimes the user don't have a lot of control of the source of data. For most cases being freed or `null` is pretty much the same, so while the error is somewhat useful in some situations, it does not really exclude argument validation.

I do agree that it should be changed in a few other places to be consistent with this if needed.

vnen commented:
Updated this to remove the check in lambda captures, since that is not needed anymore. I also added a change to allow `is` to be performed on a freed object, which will always return `false` since the type cannot be checked.

I considered doing the same for casting (`as`) but I think it's better to keep it an error. If you're casting a freed object you're most likely doing something wrong.

Also added a few test cases to validate the behavior.

Macksaur commented:
I disagree with `as` staying an error. `is`/`as` are supposed to be non-throwing and safe operators to write type-safe code, not randomly crash to the editor operators. I was very surprised they worked like this.

> If you're casting a freed object you're most likely doing something wrong.

Can you describe which scenarios this helps?

Assuming you deliberately have a freed object you cannot store/access/move it safely elsewhere or type-check with the `is`/`as` operators since it will break, despite the logic being functionally correct otherwise. With the current design you are burdened to micromanage every resource intricately. It basically makes freed objects radioactive, when they could be graceful.

I think that *all* of the VM breaks should be removed and replaced with gentle linting (that can be toggled off) rather than hard errors that prevent the user from progressing with simpler code.

romgerman commented:
> I disagree with `as` staying an error. `is`/`as` are supposed to be non-throwing and safe operators to write type-safe code, not randomly crash to the editor operators. I was very surprised they worked like this.

I agree. In my project (4.4) I must do an object validation check before `is` operator otherwise I get an error.
```gdscript
if is_instance_valid(dmg.source) and dmg.source is Player:
	_play_hit_sfx()
```
This can happen if the `dmg.source` value is a node (as it should be in my project) and the node is freed. I expect the `is` operator to work as a guard to check that because logically if an object is freed then its `null` and `null` is not of any class type. (Correct me if I'm wrong. I assume you cannot use freed objects to access their properties and methods. But maybe it's a wrong assumption)
